microdnf/commands/clean/clean.cpp

```diff
@@ -0,0 +1,145 @@
+     static_cast<CleanCommand::Actions>(CleanCommand::CLEAN_METADATA | CleanCommand::CLEAN_DBCACHE)},
+    {"dbcache", "Delete dbcache from the repositories cache", CleanCommand::CLEAN_DBCACHE},
+    {"expire-cache", "Mark the repositories cache as expired", CleanCommand::EXPIRE_CACHE}};
+
+}  // namespace
+
+
+using namespace libdnf::cli;
+
+
+CleanCommand::CleanCommand(Command & parent) : Command(parent, "clean") {
+    auto & ctx = static_cast<Context &>(get_session());
+    auto & parser = ctx.get_argument_parser();
+
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_short_description("Remove or expire cached data");
+
+    auto cache_types =
+        parser.add_new_positional_arg("cache_types", ArgumentParser::PositionalArg::AT_LEAST_ONE, nullptr, nullptr);
+    cache_types->set_short_description("List of cache types to clean up");
+
+    cache_types->set_parse_hook_func(
+        [this]([[maybe_unused]] ArgumentParser::PositionalArg * arg, int argc, const char * const argv[]) {
+            for (int i = 0; i < argc; ++i) {
+                const std::string_view cache_type{argv[i]};
+                bool found{false};
+                for (const auto & type : CACHE_TYPES) {
+                    if (cache_type == type.name) {
+                        required_actions = static_cast<Actions>(required_actions | type.action);
+                        found = true;
+                        break;
+                    }
+                }
+                if (!found) {
+                    throw std::runtime_error(fmt::format("Unknown cache type \"{}\"", argv[i]));
+                }
+            }
+            return true;
+        });
+
+    cache_types->set_complete_hook_func([](const char * arg) {
+        const std::string_view to_complete{arg};
+        std::string last;
+        std::vector<std::string> cache_types;
+        for (const auto & type : CACHE_TYPES) {
+            if (type.name.compare(0, to_complete.length(), to_complete) == 0) {
+                cache_types.emplace_back(fmt::format("{:15} ({})", type.name, type.description));
+                last = type.name;
+            }
+        }
+        if (cache_types.size() == 1) {
+            cache_types[0] = last + " ";
+        }
+        return cache_types;
+    });
+
+    cmd.register_positional_arg(cache_types);
+}
+
+
+void CleanCommand::run() {
+    auto & ctx = static_cast<Context &>(get_session());
+    fs::path cachedir{ctx.base.get_config().cachedir().get_value()};
+
+    libdnf::repo::RepoCache::RemoveStatistics statistics{};
+    for (const auto & dir_entry : std::filesystem::directory_iterator(cachedir)) {
+        libdnf::repo::RepoCache cache(ctx.base.get_weak_ptr(), dir_entry.path());
+
+        if ((required_actions & EXPIRE_CACHE) && !(required_actions & CLEAN_ALL)) {
+            cache.write_attribute(libdnf::repo::RepoCache::ATTRIBUTE_EXPIRED);
+        }
+        if (required_actions & CLEAN_ALL) {
```

Comment:
What about to test for `CLEAN_ALL` in the first place and then do not perform other clean actions? I believe that it cleans everything.

Reply:
OK. Good point.